control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This must be included before any of our internal headers because ir/std.h
+// places a vector class in the global namespace, which breaks protobuf.
+#include "p4/config/p4info.pb.h"
+
+#include <boost/optional.hpp>
+#include <boost/variant.hpp>
+#include <google/protobuf/util/json_util.h>
+#include <iostream>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/coreLibrary.h"
+#include "frontends/p4/fromv1.0/v1model.h"
+#include "frontends/p4/methodInstance.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+#include "lib/ordered_set.h"
+#include "PI/pi_base.h"
+
+#include "piSerializer.h"
+
+namespace P4 {
+namespace ControlPlaneAPI {
+
+// XXX(seth): Here are the known issues:
+// - Direct counters are not yet supported.
+// - Direct meters are not yet supported.
+// - Action profiles are not yet supported.
+// - The @id pragma is not supported for header instances and especially not for
+//   header stack instances, which currently requires allowing the programmer to
+//   provide a sequence of ids in the @id pragma. Note that @id *is* supported
+//   for header instance *fields*; this is just about the instances themselves.
+// - Locals are intentionally not exposed, but this prevents table match keys
+//   which involve complex expressions from working, because those expressions
+//   are desugared into a match against a local. This could be fixed just by
+//   exposing locals, but first we need to decide how it makes sense to expose
+//   this information to the control plane.
+// - Table match keys which involve header stacks are not supported.
+// - There is some hardcoded stuff which is tied to BMV2 and won't necessarily
+//   suit other backends. These can all be dealt with on a case-by-case basis,
+//   but they require work outside of this code.
+
+using MatchField = ::p4::config::MatchField;
+using MatchType = ::p4::config::MatchField::MatchType;
+
+/// @return true if @type has an @metadata annotation.
+static bool isMetadataType(const IR::Type* type) {
```

Comment:
The word "standard_metadata" is part of the P4 v1.0 spec, and it is hardwired in the back-end in several places. It is also available through v1model.h, and reserved (via a comment) in v1model.p4. If your code is specific to BMv2 I don't think there's a need to have a @metadata annotation.

The problem with the annotation is that it needs to be documented someplace, and it also polutes the global annotation namespace. Where is the annotation supposed to be used in P4 programs, and what does it mean? Who sets this annotation?